spot-vaults/contracts/BillBroker.sol

```diff
@@ -0,0 +1,753 @@
+        _pause();
+    }
+
+    /// @notice Unpauses deposits, withdrawals and rollovers.
+    /// @dev ERC-20 functions, like transfers will always remain operational.
+    function unpause() external onlyKeeper {
+        _unpause();
+    }
+
+    //--------------------------------------------------------------------------
+    // External & Public write methods
+
+    /// @notice Deposits usd tokens and perp tokens and mint LP tokens.
+    /// @param usdAmtMax The amount of usd tokens maximum available to be deposited.
+    /// @param perpAmtMax The amount of perp tokens maximum available to be deposited.
+    /// @param usdAmtMin The minimum amount of usd tokens that are expected to be deposited.
+    /// @param perpAmtMin The minimum amount of perp tokens that are expected to be deposited.
+    /// @return The amount of LP tokens minted.
+    function deposit(
+        uint256 usdAmtMax,
+        uint256 perpAmtMax,
+        uint256 usdAmtMin,
+        uint256 perpAmtMin
+    ) external nonReentrant whenNotPaused returns (uint256) {
+        (uint256 mintAmt, uint256 usdAmtIn, uint256 perpAmtIn) = computeMintAmt(
+            usdAmtMax,
+            perpAmtMax
+        );
+        if (mintAmt <= 0) {
+            return 0;
+        }
+        if (usdAmtIn < usdAmtMin || perpAmtIn < perpAmtMin) {
+            revert SlippageTooHigh();
+        }
+
+        // Transfer perp and usd tokens from the user
+        usd.safeTransferFrom(_msgSender(), address(this), usdAmtIn);
+        perp.safeTransferFrom(_msgSender(), address(this), perpAmtIn);
+
+        // mint LP tokens
+        _mint(_msgSender(), mintAmt);
+        return mintAmt;
+    }
+
+    /// @notice Burns LP tokens and redeems usd and perp tokens.
+    /// @param burnAmt The LP tokens to be burnt.
+    /// @return usdAmtOut The amount usd tokens returned.
+    /// @return perpAmtOut The amount perp tokens returned.
+    function redeem(
+        uint256 burnAmt
+    ) external nonReentrant whenNotPaused returns (uint256, uint256) {
+        if (burnAmt <= 0) {
+            return (0, 0);
+        }
+
+        (uint256 usdAmtOut, uint256 perpAmtOut) = computeRedemptionAmts(burnAmt);
+
+        // burn LP tokens
+        _burn(_msgSender(), burnAmt);
+
+        // return funds
+        usd.safeTransfer(_msgSender(), usdAmtOut);
+        perp.safeTransfer(_msgSender(), perpAmtOut);
+        return (usdAmtOut, perpAmtOut);
+    }
+
+    /// @notice Swaps usd tokens from the user for perp tokens from the reserve.
+    /// @param usdAmtIn The amount of usd tokens swapped in.
+    /// @param perpAmtMin The minimum amount of perp tokens that are expected out.
+    /// @return perpAmtOut The amount perp tokens swapped out.
+    function swapUSDForPerps(
+        uint256 usdAmtIn,
+        uint256 perpAmtMin
+    ) external nonReentrant whenNotPaused returns (uint256) {
+        // Transfer usd tokens from user
+        usd.safeTransferFrom(_msgSender(), address(this), usdAmtIn);
```

Comment:
Hmm, I guess there's an issue here. Because the `reserveState` call right after this assumes that the transfer has not happened; so we're double counting the incoming assets.  Moving the transfer to after the `compute`